### ISSUE TITLE
Allow choosing position of JS and CSS in template with render tag.

### ISSUE THREAD

**Issue by sammcqueen** [verbb/comments#177]
Using the default template tags `{{ craft.comments.render(entry.id) }}`, the JS is always included at the end of the template.

It would be great to be able to use the out of the box JS, CSS and templates, _but_ give the user the option to include it wherever they want in their template.

Perhaps with seperate `{{ craft.comments.renderCss }}` and `{{ craft.comments.renderJs }}` tags?

This is especially useful for sites with ajax page loads or single page web apps where only a certain portion of the page might be loaded at a time. If this is the case the JS and CSS never gets injected.

This could be done by manually included the prebuilt JS but unfortunately many of the settings are handed to it from [CommentsService.php](https://github.com/verbb/comments/blob/craft-3/src/services/CommentsService.php#L85)

A completely custom JS solution could be built, but all of the required JS and functionality is already there. Hope that makes sense!

**Comment by engram-design** [comment]
I'll have to do some investigation with Yii's Asset Bundles. There doesn't seem to be a great way to handle this, as Craft and Yii2 register CSS/JS to a View class, which takes care of the actual rendering. I know I can create functions to look at a JS/CSS file and just spit out the content, but I'm not sure how I feel about that robustness.

**Comment by john-henry** [comment]
I would love this too as finding it impossible to use with [Blitz plugin](https://github.com/putyourlightson/craft-blitz) right now. I've had to disable my comments section as I'm injecting the comments into a page via a Blitz include and of course the JS is never injected. The CSS is no problem as I can compile that myself.

**Comment by engram-design** [comment]
Added in [1.6.3](https://github.com/verbb/comments/releases/tag/1.6.3). Refer to [docs](https://verbb.io/craft-plugins/comments/docs/template-guides/available-variables)

**Comment by davecoggins** [comment]
I'm currently trying to achieve the same thing with blitz, I've updated comments to 1.6.4 and have the following in my blog post template:

`{{ craft.blitz.getTemplate('_dynamic/comments', { id: entry.id }) }}`

and then in _dynamic/comments file I have:

```
{{craft.comments.renderCss(id)}}  
{{ craft.comments.render(id) }}
{{craft.comments.renderJs(id)}}
```

The comment block is injected as expected along with css and js but when I try and leave comment the js doesn't seem to be hooking up to the submit button. Instead the form submits and causing a page refresh. No comments are recorded in the CMS.

Looking through the log files it looks like it is being marked as spam:

```
Element not saved due to validation error: Array
(
    [comment] => Array
        (
            [0] => Oops, comment validation failed. Marked as spam.
        )

)
```

If I toggle off extra spam protection in settings the comment does get saved to the database, on the front end this still triggers a page refresh on submit.

I am going about this the wrong way?

Thanks in advance

**Comment by john-henry** [comment]
Hi @davecoggins 

Working with the same issue myself and its not necessarily a Comments plugin issue. If you put renderJs inside the Blitz template then it's not going to run anything inside the <script> tags injected as per the Blitz docs

If you do this
```
{{ craft.blitz.getTemplate('_dynamic/comments', { id: entry.id }) }}
{{craft.comments.renderJs(id)}}
```
Then you can see it working.

My issue is that I'm using Cloudflare on top so this causes further issues.

I think we should be posting over on the Blitz plugin GH issues to come up with a solution for this one as the Comments plugin has given us all the bits we need so far

**Comment by davecoggins** [comment]
Hi @john-henry 

Thanks for you reply, I've tried what you suggested but I get a js error:

`comments.js?v=1592899368:1 Uncaught TypeError: Cannot read property 'querySelectorAll' of null`

I think this is due to the fact the javascript executes before the blitz dynamic content has arrived.

I think the solution is to use one of the blitz hooks:

```
// The `afterBlitzInjectAll` event is triggered after all dynamic content tag is injected
document.addEventListener('afterBlitzInjectAll', function (event) {
								
});
```

To wrap the comment instantiation code so it is only executed once the html has arrived. I'm not sure the best way to go about this but one idea would be to separate out the render of comments.js file and the specific instantiation code and for that code not to be wrapped in script tags so that it could be placed in the above callback function.

But maybe there is a better way?

**Comment by tomkiss** [comment]
Hey @davecoggins,

I also had the error with `Cannot read property 'querySelectorAll' of null` when using Blitz and Comments. My part-solution was to use the [craft.comments.getJsVariables](https://verbb.io/craft-plugins/comments/docs/template-guides/available-variables#craft-comments-getJsVariables) example like so:

```
  {# Blitz-compatible setup #}
  {{ craft.blitz.getTemplate('_dynamic/comments', { id: entry.id }) }}
  
  {# Render the `comments.js` file - exclude the inline JS #}
  {{ craft.comments.renderJs(entry.id, {}, false) }}

  {# Fetch the variables required for the comments JS #}
  {%- set jsVariables = craft.comments.getJsVariables(entry.id) %}
  
  {# Initialise once blitz event is triggered #}
  <script type="text/javascript">
    document.addEventListener('afterBlitzInjectAll', function (event) {
      new Comments.Instance('{{ jsVariables.id }}', {{ jsVariables.settings | json_encode | raw }});                      
    });
  </script>
```

This all works... until I enable reCaptcha, which then gives a `Uncaught ReferenceError: grecaptcha is not defined` console error.

Did you experience the reCaptcha error or find any other solutions at all?

**Comment by john-henry** [comment]
@tomkiss Your reply alerted to me of getJsVariables now available in 1.6.6. Using your method I now have working comments, I don't use reCaptcha however so can't help you there, unfortunately.

**Comment by john-henry** [comment]
This actually requires a bit more work.  The above will just cache a CSRF token for the form and you will run into issues.

This is what I did. I  copied the _special templates from the vendor folder into my own templates folder. I removed the `{{ csrfInput() }}` tag in `form.html`

In my entry template, I did the following

```
		{{ craft.blitz.getTemplate('/_partials/comments',{entryId: entry.id}) }}

		{# Render the `comments.js` file - exclude the inline JS #}
		{{ craft.comments.renderJs(entry.id, {}, false) }}

		{# Fetch the variables required for the comments JS #}
		{% set jsVariables = craft.comments.getJsVariables(entry.id) %}

		<script type="text/javascript">
                       // Wait for the document to be ready, then initialise
			document.addEventListener('afterBlitzInjectAll', function (event) {

			// Use Blitz controller action to set a dynamic csrfToken https://putyourlightson.com/plugins/blitz#csrf-controller-actions
			fetch('/actions/blitz/csrf/token')
				.then(result => { return result.text(); })
				.then(result => {

					// Turn twig jsVariables.settings array into JS array
					let jsVariables = {{ jsVariables.settings | json_encode | raw }};

					// Merge the new dynamic csrfToken into our jsVariables array
					jsVariables["csrfToken"] = result;

					
							// Create new Comments instance using our updated jsVariables array
							new Comments.Instance('{{ jsVariables.id }}', jsVariables);
					
				});

});
		</script>
```
This will give you a dynamic CSRF in every page load and inject it into the Comments instance.
